### PR TITLE
Padding & Anchor Overhaul, Cullling, System refactor.

### PR DESCRIPTION
**Pull request by TheGrimsey**
**Changes:**
- Update how we handle positioning of elements to allow anchoring to parts of the screen and not just to parts of other elements.
- Add padding for elements.
- Clean up some includes.
- Add util function for converting from window coordinate space to ui coordinate space.
- Refactor many names for consistency.
- Add some missing functions to elements.
- Split off children & parent from Transform to Relation component.
- Add Root component to mark elements as top of the parent tree.
- Refactor element sorting. *Twice*
- Split UpdateElementSystem into multiple parallel ones.
- Cull off screen elements.
- Add a UI stats tab to Engine Info.

**Tests Performed:**

- [X] Build
- [X] Tested in-game

**Targeted branch(es):**
- [x] Master

### CODE REVIEW

File: client/UI/ECS/Systems/BuildSortKeySystem.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "../../Utils/SortUtils.h"
+
+namespace UISystem
+{
+    void BuildSortKeySystem::Update(entt::registry& registry)
+    {
+        auto sortView = registry.view<UIComponent::Relation, UIComponent::SortKey, UIComponent::SortKeyDirty>();
+        sortView.each([&](entt::entity entity, UIComponent::Relation& relation, UIComponent::SortKey& sortKey)
+            {
```
Comment (Pursche) [inline client/UI/ECS/Systems/BuildSortKeySystem.cpp:17]: Indentation issue

File: client/UI/ECS/Systems/DeleteElementsSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace UISystem
+{
+    void DeleteElementsSystem::Update(entt::registry& registry)
+    {
+        auto& dataSingleton = registry.ctx<UISingleton::UIDataSingleton>();
+
+        // Destroy elements queued for destruction.
+        auto deleteView = registry.view<UIComponent::Destroy>();
+        deleteView.each([&](entt::entity entityId) {
```
Comment (Pursche) [inline client/UI/ECS/Systems/DeleteElementsSystem.cpp:16]: Curly brace on new line and double check indentation please

File: client/UI/ECS/Systems/UpdateBoundsSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#include "../../Utils/ColllisionUtils.h"
+
+
+namespace UISystem
+{
+    void UpdateBoundsSystem::Update(entt::registry& registry)
+    {
+        auto boundsUpdateView = registry.view<UIComponent::Transform, UIComponent::Collision, UIComponent::Relation, UIComponent::BoundsDirty>();
+        boundsUpdateView.each([&](entt::entity entityId, UIComponent::Transform& transform, UIComponent::Collision& collision, UIComponent::Relation& relation)
+            {
```
Comment (Pursche) [inline client/UI/ECS/Systems/UpdateBoundsSystem.cpp:17]: Fix indentation issue

File: client/UI/ECS/Systems/UpdateCullingSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        auto oldCulledView = registry.view<UIComponent::NotCulled, UIComponent::Dirty>();
+        registry.remove<UIComponent::NotCulled>(oldCulledView.begin(), oldCulledView.end());
+
+        auto cullView = registry.view<UIComponent::Transform, UIComponent::Dirty>();
+        std::vector<entt::entity> notCulled;
+        notCulled.reserve(cullView.size());
+
+        cullView.each([&](entt::entity entity, UIComponent::Transform& transform)
+            {
```
Comment (Pursche) [inline client/UI/ECS/Systems/UpdateCullingSystem.cpp:26]: Fix indentation issue

File: client/UI/ECS/Systems/UpdateRenderingSystem.cpp [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    void UpdateRenderingSystem::Update(entt::registry& registry)
+    {
+        Renderer::Renderer* renderer = ServiceLocator::GetRenderer();
+        auto& dataSingleton = registry.ctx<UISingleton::UIDataSingleton>();
+
+        auto inputFieldView = registry.view<UIComponent::Transform, UIComponent::InputField, UIComponent::Text, UIComponent::Dirty>();
+        inputFieldView.each([&](const UIComponent::Transform& transform, const UIComponent::InputField& inputField, UIComponent::Text& text)
+            {
```
Comment (Pursche) [inline client/UI/ECS/Systems/UpdateRenderingSystem.cpp:63]: Fix indentation please

File: client/UI/ECS/Systems/UpdateRenderingSystem.cpp [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                text.pushback = UIUtils::Text::CalculatePushback(&text, inputField.writeHeadIndex, 0.2f, transform.size.x, transform.size.y);
+            });
+
+        auto imageView = registry.view<UIComponent::Transform, UIComponent::Image, UIComponent::Dirty>();
+        imageView.each([&](UIComponent::Transform& transform, UIComponent::Image& image)
+            {
```
Comment (Pursche) [inline client/UI/ECS/Systems/UpdateRenderingSystem.cpp:69]: Fix indentation please

File: client/UI/ECS/Systems/UpdateRenderingSystem.cpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+                }
+
+                void* dst = renderer->MapBuffer(image.vertexBufferID);
+                memcpy(dst, vertices.data(), bufferSize);
+                renderer->UnmapBuffer(image.vertexBufferID);
+            });
+
+        auto textView = registry.view<UIComponent::Transform, UIComponent::Text, UIComponent::Dirty>();
+        textView.each([&](UIComponent::Transform& transform, UIComponent::Text& text)
+            {
```
Comment (Pursche) [inline client/UI/ECS/Systems/UpdateRenderingSystem.cpp:127]: Fix indentation please

**Review by Pursche — CHANGES_REQUESTED** [review]
Looks good, mostly indentation issues.

**Comment by TheGrimsey** [comment]
Should all be fixed now

**Review by Pursche — APPROVED** [review]

**Pull request merged**